client/lib/interval/runner.js

```diff
@@ -0,0 +1,97 @@
+import { fromJS } from 'immutable';
+
+export const EVERY_SECOND = 1000;
+export const EVERY_FIVE_SECONDS = 5 * 1000;
+export const EVERY_TEN_SECONDS = 10 * 1000;
+export const EVERY_THIRTY_SECONDS = 30 * 1000;
+export const EVERY_MINUTE = 60 * 1000;
+
+const initialState = fromJS( {
+	nextId: 1,
+	periodTimers: {
+		EVERY_SECOND: null,
+		EVERY_FIVE_SECONDS: null,
+		EVERY_TEN_SECONDS: null,
+		EVERY_THIRTY_SECONDS: null,
+		EVERY_MINUTE: null
+	},
+	actions: []
+} );
+let state = initialState;
+
+const increment = a => a + 1;
+const push = item => list => list.push( item );
+const pull = id => list => list.filterNot( o => o.get( 'id' ) === id );
+
+export const resetForTesting = () => {
+	state
+		.get( 'periodTimers' )
+		.forEach( clearTimeout );
+
+	state = initialState;
+};
+
+export function add( period, onTick ) {
+	const id = state.get( 'nextId' );
+
+	storeNewAction( { id, period, onTick } );
+	scheduleNextRun();
+
+	return id;
+}
+
+function storeNewAction( { id, period, onTick } ) {
+	state = state
+		.update( 'actions', push( fromJS( { id, period, onTick } ) ) )
```

Comment:
The `immutable.js` docs are hard to navigate at first, I think. Probably the main reason is that almost everything in that library stems from `Utterable` or `KeyedIterable` base classes. In this code you figured it all out properly, but the piece that would have helped is that this is a [`Map`](http://facebook.github.io/immutable-js/docs/#/Map/update). `update` is one of a few basic ways to set values in the data structure. It differs from `set` in that instead of just supplying a key and a value, you supply a key and it provides the existing value as input into a function you pass. The output of that function gets returned as the new value for the supplied key.

I will note this here - when I first began with `Immutable.js` it seemed very strange to me. However, once I figured out the basics of using it I found it to be refreshingly easy and simple, especially in contrast to using JavaScript's home-built immutability operations.